packages/nimble-components/src/wafer-map/specs/features/rendering.md

```diff
@@ -0,0 +1,207 @@
+
+| FTL      | die count | size     |
+| -------- | --------- | -------- |
+| baseline | 13 dies   | 49.78MB  |
+|          | 1M dies   | 509.81MB |
+| 1 thread | 13 dies   | 40.40MB  |
+|          | 1M dies   | 44.65MB  |
+| 2 thread | 13 dies   | 40.50MB  |
+|          | 1M dies   | 44.56MB  |
+
+Using TypedArrays seems to lower the memory needed for large data sets.
+The arrays are cloned before being sent to the threads, but this does not seem to impact the needed memory very much.
+
+There are also no noticeable degradations in user experience or interactions with smaller data sets.
+
+## Links To Relevant Work Items and Reference Material
+
+[Feature 2391160: Faster Rendering of Large Wafer Maps](https://dev.azure.com/ni/DevCentral/_workitems/edit/2391160)
+
+[Faster Rendering of Large Wafer Maps HLD](https://ni.visualstudio.com/DevCentral/_git/Skyline?path=/docs/design-documents/Ozone/Requirements/Optimize-Wafer-Map-to-Handle-NXP-s-340k-die-per-wafer-data-set/Faster-rendering-of-Large-Wafer-Maps.md&version=GC00f8bb9e698a7310f68fa54395eee63a99cec368&_a=preview)
+
+[Worker Rendering POC](https://github.com/ni/nimble/compare/main...users/munteannatan/scale_and_test_updates)
+
+## Implementation / Design
+
+We will change the whole process of ingesting the data and rendering it in the canvas. The main changes are the data structures used and the process from single threaded to multi threaded. We expect that using better structured data will lead to time improvements when iterating over values which will improve the rendering and allow distributing the workload if necessary. Moreover, the adoption of thread based processing will free the main thread of the browser to handle user requests and other page events, while allowing us to scale in the future by parallel processing if the need arises.
+
+The detailed plan can be found in the azure HLD [Faster Rendering of Large Wafer Maps HLD](https://ni.visualstudio.com/DevCentral/_git/Skyline?path=/docs/design-documents/Ozone/Requirements/Optimize-Wafer-Map-to-Handle-NXP-s-340k-die-per-wafer-data-set/Faster-rendering-of-Large-Wafer-Maps.md&version=GC00f8bb9e698a7310f68fa54395eee63a99cec368&_a=preview).
+
+The POC is found in this branch [Worker Rendering POC](https://github.com/ni/nimble/compare/main...users/munteannatan/scale_and_test_updates).
+
+### Data Structure and Interface
+
+This is the proposed representation of the wafer information in memory
+
+```TS
+class WaferData {
+    // the x coordinates of each column of dies
+    dieColIndexArray: Int32Array;
+    // the lengths of each row of dies
+    rowLengthsArray: Int32Array;
+    // the y coordinates of each die as a matrix row by row
+    dieRowIndexLayer: Int32Array;
+    // the value of each die as a matrix row by row
+    dieValuesLayer: Int32Array;
+    // the highlight state of each die as a matrix row by row
+    dieHighlightsLayer: Int8Array;
```

Comment:
There are a couple potential approaches that have not been decided yet and are open issues if this data format is used:

1. highlights are treated as an array of booleans as UInt8Array. That would be a bigger change from the public api and require users to pre-calculate tags into highlighted conditions.
2. Treat the tags as a bitset as a Int32Array. We don't care about the tag strings we just keep a 32-bitset of tags associated with a wafer. This more closely aligns to the public api but limits you to 32 tags (or whatever size is chosen for the array)